flaky test explaintest

## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

<!-- a step by step guide for reproducing the bug. -->

### 2. What did you expect to see? (Required)

### 3. What did you see instead (Required)

```
[2023/07/31 14:05:45.699 +08:00] [ERROR] [main.go:674] ["open Db failed"] ["take time"=30.066849144s] [error="dial tcp [::1]:35611: connect: connection refused"]

[2023/07/31 14:05:45.700 +08:00] [FATAL] [main.go:694] ["open DB failed"] [error="dial tcp [::1]:35611: connect: connection refused"] [errorVerbose="dial tcp [::1]:35611: connect: connection refused\ngithub.com/pingcap/errors.AddStack\n\t/go/pkg/mod/github.com/pingcap/errors@v0.11.5-0.20221009092201-b66cddb77c32/errors.go:174\ngithub.com/pingcap/errors.Trace\n\t/go/pkg/mod/github.com/pingcap/errors@v0.11.5-0.20221009092201-b66cddb77c32/juju_adaptor.go:15\nmain.openDBWithRetry\n\t/home/jenkins/agent/workspace/pingcap/tidb/ghpr_check/tidb/cmd/explaintest/main.go:675\nmain.main\n\t/home/jenkins/agent/workspace/pingcap/tidb/ghpr_check/tidb/cmd/explaintest/main.go:689\nruntime.main\n\t/usr/local/go/src/runtime/proc.go:250\nruntime.goexit\n\t/usr/local/go/src/runtime/asm_amd64.s:1598"] [stack="main.main\n\t/home/jenkins/agent/workspace/pingcap/tidb/ghpr_check/tidb/cmd/explaintest/main.go:694\nruntime.main\n\t/usr/local/go/src/runtime/proc.go:250"]

```
### 4. What is your TiDB version? (Required)

<!-- Paste the output of SELECT tidb_version() -->

--- a/cmd/explaintest/BUILD.bazel
+++ b/cmd/explaintest/BUILD.bazel
@@ -10,6 +10,7 @@ go_library(
         "//session",
         "//sessionctx",
         "//util/logutil",
+        "//util/mathutil",
         "//util/mock",
         "@com_github_go_sql_driver_mysql//:mysql",
         "@com_github_pingcap_errors//:errors",

--- a/cmd/explaintest/main.go
+++ b/cmd/explaintest/main.go
@@ -33,6 +33,7 @@ import (
 	"github.com/pingcap/tidb/session"
 	"github.com/pingcap/tidb/sessionctx"
 	"github.com/pingcap/tidb/util/logutil"
+	"github.com/pingcap/tidb/util/mathutil"
 	"github.com/pingcap/tidb/util/mock"
 	"go.uber.org/zap"
 )
@@ -649,14 +650,19 @@ func loadAllTests() ([]string, error) {
 // openDBWithRetry opens a database specified by its database driver name and a
 // driver-specific data source name. And it will do some retries if the connection fails.
 func openDBWithRetry(driverName, dataSourceName string) (mdb *sql.DB, err error) {
+	now := time.Now()
 	startTime := time.Now()
-	sleepTime := time.Millisecond * 500
-	retryCnt := 60
-	// The max retry interval is 30 s.
-	for i := 0; i < retryCnt; i++ {
+	sleepTime := time.Millisecond * 100
+	maxSleepTime := time.Second * 2
+	// The max retry interval is 60 s.
+	maxDuration := time.Minute
+	for {
+		if time.Since(now) > maxDuration {
+			break
+		}
 		mdb, err = sql.Open(driverName, dataSourceName)
 		if err != nil {
-			log.Warn("open DB failed", zap.Int("retry count", i), zap.Error(err))
+			log.Warn("open DB failed", zap.Duration("retry duration", time.Since(now)), zap.Error(err))
 			time.Sleep(sleepTime)
 			continue
 		}
@@ -664,11 +670,13 @@ func openDBWithRetry(driverName, dataSourceName string) (mdb *sql.DB, err error)
 		if err == nil {
 			break
 		}
-		log.Warn("ping DB failed", zap.Int("retry count", i), zap.Error(err))
+		log.Warn("ping DB failed", zap.Duration("retry duration", time.Since(now)), zap.Error(err))
 		if err1 := mdb.Close(); err1 != nil {
 			log.Error("close DB failed", zap.Error(err1))
 		}
 		time.Sleep(sleepTime)
+		sleepTime = sleepTime * 2
+		sleepTime = mathutil.Min(maxSleepTime, sleepTime)
 	}
 	if err != nil {
 		log.Error("open Db failed", zap.Duration("take time", time.Since(startTime)), zap.Error(err))

--- a/cmd/explaintest/main.go
+++ b/cmd/explaintest/main.go
@@ -660,6 +660,9 @@ func openDBWithRetry(driverName, dataSourceName string) (mdb *sql.DB, err error)
 		if time.Since(now) > maxDuration {
 			break
 		}
+		time.Sleep(sleepTime)
+		sleepTime = sleepTime * 2
+		sleepTime = mathutil.Min(maxSleepTime, sleepTime)
 		mdb, err = sql.Open(driverName, dataSourceName)
 		if err != nil {
 			log.Warn("open DB failed", zap.Duration("retry duration", time.Since(now)), zap.Error(err))
@@ -674,9 +677,6 @@ func openDBWithRetry(driverName, dataSourceName string) (mdb *sql.DB, err error)
 		if err1 := mdb.Close(); err1 != nil {
 			log.Error("close DB failed", zap.Error(err1))
 		}
-		time.Sleep(sleepTime)
-		sleepTime = sleepTime * 2
-		sleepTime = mathutil.Min(maxSleepTime, sleepTime)
 	}
 	if err != nil {
 		log.Error("open Db failed", zap.Duration("take time", time.Since(startTime)), zap.Error(err))

